Switch to enable Marionette in Firefox

vkatsikaros opened the pull request:
Hi

This patch adds a switch when launching Firefox, to setup and enable the Marionette server (all latest Firefox binary already come with it built in). The default Marionette port is used, or if already used a higher available one is used.

```
use Selenium::Firefox;
my $driver = Selenium::Firefox->new( marionette_enabled => 1 );
```

A Marionette client (or the wires proxy server) is still needed in order to communicate with the browser's Marionette server (this patch doesn't add support for the Marionette protocol). However, this is still useful as Firefox's configuration lives in one place.

Hope this is useful for general purpose. Happy to discuss any changes if needed.

Regards
Vangelis

vkatsikaros commented:
Ah my bad :( didn't run the tests beforehand and didn't catch
```
t/CanStartBinary.t ...................... Can't locate object method "marionette_enabled" via package "Selenium::PhantomJS" at /home/travis/build/gempesaw/Selenium-Remote-Driver/.build/oQazfqPHTE/blib/lib/Selenium/CanStartBinary.pm line 128.
```

gempesaw commented on lib/Selenium/Firefox.pm:
```diff
@@ -84,6 +88,37 @@ has '+wd_context_prefix' => (
+    default => sub { 2828 }
+);
+
+=attr marionette_binary_port
```
oops, looks like this line should say `marionette_enabled`

vkatsikaros replied:
oops! fixed in 8d4ac58

gempesaw commented on lib/Selenium/Remote/Driver.pm:
```diff
@@ -144,6 +144,7 @@ you please.
         'javascript'           - <boolean>  - whether javascript should be supported
         'accept_ssl_certs'     - <boolean>  - whether SSL certs should be accepted, default is true.
         'firefox_profile'      - Profile    - Use Selenium::Firefox::Profile to create a Firefox profile for the browser to use
+        'marionette_enabled'   - <boolean>  - whether Firefox should enable Marionette. default if false
```
These options are for the Selenium::Remote::Driver which doesn't do the binary start up, so I don't think this option belongs here. Unless you mean that passing `marionette_enabled` to a running `selenium-standalone-server-x.xx.x.jar` would start marionette mode, in which case, never mind :D

vkatsikaros replied:
Ah! I thought these were all the available options, that's why I added it there. No I haven't ever used the selenium standalone server.

I'll probably remove it from the docs here.

gempesaw commented on lib/Selenium/CanStartBinary.pm:
```diff
@@ -121,6 +121,20 @@ has '+port' => (
+    builder => sub {
+        my ($self) = @_;
+
+        if ($self->isa('Selenium::Firefox') && $self->marionette_enabled) {
```
off the cuff: does it make sense/is it possible to move this logic in to the `Selenium::Firefox` class itself to avoid the `$self->isa` check entirely? seems _slightly_ cleaner

gempesaw commented:
Awesome, thanks! I've looked over the code a bit but still haven't had a chance to run it locally.

If it's running in marionette mode, does it still use the `webdriver_port` ? I'm not familiar with how the marionette mode works. If they're not both used at the same time, does it make sense to re-use the existing `binary_port` and `port` attributes instead of making a separate set? Or does that actually make things more complicated.

vkatsikaros commented:
Thanks for taking a look. Yes it still uses the `webdriver_port` as the protocol used to talk to marionette is different. 

With `marionette` you can interact with the webpage and the browser's chrome (example change the configuration of a *running* browser), and with `webdriver` you can interact only with the webpage. For convenience and to ensure a smoother future transition (see below), I use `marionette` to interact with the browser and `webdriver` to interact with the webpage.

In the future, it seems that  [`Marionette, the next generation of FirefoxDriver`](https://developer.mozilla.org/en-US/docs/Mozilla/QA/Marionette/WebDriver) will replace the webdriver as it will be offering a superset of commands. However this isn't ready yet.

gempesaw commented:
Sorry for the delay, and, cool, thanks for the explanation. Looks like there are a few outstanding items:

- I'm curious if we can move the `marionette_port` to the `Selenium::Firefox` role (https://github.com/gempesaw/Selenium-Remote-Driver/pull/268#discussion_r63225923)
- Can you add an example in the POD of `Selenium::Firefox` about turning on the marionette option during construction? Either at the bottom of the SYNOPSIS or DESCRIPTION would be fine.

Thanks!

vkatsikaros commented:
Sorry for the terrible delay. I added examples for all the `=attr` this patch adds. I used the example of `startup_timeout` that has the example in the attribute documentation. I also added 1 line in SYNOPSIS.

Regarding moving the mariontte port, I remember I had an initialization problem, I'll look again and let you know.

vkatsikaros commented:
I am running the tests with:
```
$ export RELEASE_TESTING=1
$ export DISPLAY='vkcloud2:1'
$ perl t/CanStartBinary.t 
ok 1 # skip Phantomjs binary not found in path
ok 2 # skip Phantomjs binary not found in path
ok 3 # skip Phantomjs binary not found in path
ok 4 - we throw if the user specified binary is not executable
ok 5 # skip PhantomJS needed for manual binary path tests
ok 6 # skip PhantomJS needed for manual binary path tests
ok 7 # skip Chrome binary not found in path
ok 8 # skip Chrome binary not found in path
ok 9 # skip Chrome binary not found in path
ok 10 - we can find some sort of firefox
ok 11 - firefox can start up its own binary
ok 12 - the firefox binary is listening on its port
ok 13 - firefox can start up its own binary
ok 14 - the firefox binary with marionette enabled is listening on its marionette port
ok 15 - We can specify how long to wait for a binary to be available
```

gempesaw commented:
cool, thanks! and no worries about the delays, I, too, am slow. c: I think FF48 is out today/soon, so this is _quite_ timely

gempesaw commented:
@vkatsikaros hey what version of Firefox are you using? I've tried with 47 and 48 and I can't get the `CanStartBinary.t` to pass, it isn't starting up the Firefox session or something. I'll keep looking into it ....

vkatsikaros commented:
Hm, I used 46 and when it autoupdated to 47 indeed it stopped working, but thought it was a webdriver version issue and didn't investigate further. My bad, will have to check again...

gempesaw commented:
no problem, I know 47 and 48 have some issues with no longer blindly accepting unsigned extensions, good to know that it worked with 46 though. c:

gempesaw commented:
cool, I rebased this to master, and pushed it so it's in origin/master now. Github's interface doesn't seem to pick it up, so I'm closing this manually.

and then I added the geckodriver stuff in the [geckodriver branch](https://github.com/gempesaw/Selenium-Remote-Driver/tree/geckodriver).

thanks!
